hcl2template/addrs/plugin.go

```diff
@@ -13,21 +17,53 @@ import (
 
 // Plugin encapsulates a single plugin type.
 type Plugin struct {
-	Hostname  string
-	Namespace string
-	Type      string
+	Source string
 }
 
-func (p Plugin) RealRelativePath() string {
-	return p.Namespace + "/packer-plugin-" + p.Type
+// Parts returns the list of components of the source URL, starting with the
+// host, and ending with the name of the plugin (non-simplified).
+//
+// This will correspond more or less to the filesystem hierarchy where
+// the plugin is installed.
+func (p Plugin) Parts() []string {
+	var retParts []string
+
+	// This should not happen, we're supposed to check that the source isn't
+	// empty when creating the plugin, but let's be careful here and not
+	// misbehave.
+	if p.Source == "" {
+		log.Printf("[ERROR] - Plugin.Parts: the source was empty, this should not happen in normal usage, and is likely a Packer bug.")
+		return retParts
+	}
+
+	dir, part := path.Split(strings.TrimRight(p.Source, "/"))
+	for dir != "" {
+		retParts = append(retParts, part)
+		dir, part = path.Split(strings.TrimRight(dir, "/"))
+	}
+	retParts = append(retParts, part)
+
+	// The algorithm above registers the path in reversed order
+	//
+	// The function is supposed to return them in the host->...->name
+	// order however, so we have to reverse the slice.
+	slices.Reverse(retParts)
+
+	return retParts
 }
 
-func (p Plugin) Parts() []string {
-	return []string{p.Hostname, p.Namespace, p.Type}
+// Name returns the raw name of the plugin from its source
+//
+// Exemples:
+//   - "github.com/hashicorp/amazon" -> "amazon"
+//   - "github.com/hashicorp/packer-plugin-amazon" -> "amazon"
+func (p Plugin) Name() string {
+	parts := p.Parts()
+	return strings.ReplaceAll(parts[len(parts)-1], "packer-plugin-", "")
```

Comment:
I believe this still supports plugin Source addresses in the form of "github.com/hashicorp/packer-plugin-docker" right. Lets break this out into separate PR so we can validate it doesn't breaking anything given packer init original implementation.

Reply:
So theoretically the structure allows it, we could get rid of that `ReplaceAll` though.

The check is done at creation-time, in the `ParsePluginSourceString` function, there we check if the name contains `packer-` or `packer-plugin-`, and we error.
It's not bulletproof as the structure can still be created from scratch, so that protection could be circumvented (this was already the case though to be clear, it's not something we are introducing right now)